File: App/app/src/main/java/com/cinema/ticket_booking/data/repository/ShowtimeRepository.java
package com.cinema.ticket_booking.data.repository;

import androidx.lifecycle.*;
import com.cinema.ticket_booking.data.model.response.*;
import com.cinema.ticket_booking.network.ApiService;
import com.cinema.ticket_booking.util.Resource;
import retrofit2.Call; import retrofit2.Callback; import retrofit2.Response;
import java.util.List;
import javax.inject.Inject; import javax.inject.Singleton;

@Singleton
public class ShowtimeRepository {
    private final ApiService api;
    @Inject public ShowtimeRepository(ApiService api) { this.api = api; }

    public LiveData<Resource<List<ShowtimeResponse>>> getShowtimes(String movieId, String cinemaId, String date) {
        MutableLiveData<Resource<List<ShowtimeResponse>>> r = new MutableLiveData<>();
        r.setValue(Resource.loading());
        api.getShowtimes(movieId, cinemaId, date).enqueue(new Callback<>() {
            @Override public void onResponse(Call<ApiResponse<List<ShowtimeResponse>>> c, Response<ApiResponse<List<ShowtimeResponse>>> res) {
                if (res.isSuccessful() && res.body() != null && res.body().isSuccess())
                    r.setValue(Resource.success(res.body().getData()));
                else r.setValue(Resource.error("Tải suất chiếu thất bại"));
            }
            @Override public void onFailure(Call<ApiResponse<List<ShowtimeResponse>>> c, Throwable t) {
                r.setValue(Resource.error("Lỗi kết nối: " + t.getMessage()));
            }
        });
        return r;
    }

    public LiveData<Resource<SeatMapResponse>> getSeatMap(String showtimeId) {
        MutableLiveData<Resource<SeatMapResponse>> r = new MutableLiveData<>();
        r.setValue(Resource.loading());
        api.getSeatMap(showtimeId).enqueue(new Callback<>() {
            @Override public void onResponse(Call<ApiResponse<SeatMapResponse>> c, Response<ApiResponse<SeatMapResponse>> res) {
                if (res.isSuccessful() && res.body() != null && res.body().isSuccess())
                    r.setValue(Resource.success(res.body().getData()));
                else r.setValue(Resource.error("Tải sơ đồ ghế thất bại"));
            }
            @Override public void onFailure(Call<ApiResponse<SeatMapResponse>> c, Throwable t) {
                r.setValue(Resource.error("Lỗi kết nối: " + t.getMessage()));
            }
        });
        return r;
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/data/repository/ShowtimeRepository.kt
package com.cinema.ticket_booking.data.repository

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.cinema.ticket_booking.data.model.response.ApiResponse
import com.cinema.ticket_booking.data.model.response.SeatMapResponse
import com.cinema.ticket_booking.data.model.response.ShowtimeResponse
import com.cinema.ticket_booking.network.ApiService
import com.cinema.ticket_booking.util.Resource
import javax.inject.Inject
import javax.inject.Singleton
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

@Singleton
class ShowtimeRepository @Inject constructor(
    private val api: ApiService
) {

    fun getShowtimes(movieId: String?, cinemaId: String?, date: String?): LiveData<Resource<List<ShowtimeResponse>>> {
        val r = MutableLiveData<Resource<List<ShowtimeResponse>>>()
        r.value = Resource.loading()
        api.getShowtimes(movieId, cinemaId, date).enqueue(object : Callback<ApiResponse<List<ShowtimeResponse>>> {
            override fun onResponse(
                call: Call<ApiResponse<List<ShowtimeResponse>>>,
                res: Response<ApiResponse<List<ShowtimeResponse>>>
            ) {
                if (res.isSuccessful && res.body() != null && res.body()!!.success) {
                    r.value = Resource.success(res.body()!!.data)
                } else {
                    r.value = Resource.error("Tải suất chiếu thất bại")
                }
            }

            override fun onFailure(call: Call<ApiResponse<List<ShowtimeResponse>>>, t: Throwable) {
                r.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return r
    }

    fun getSeatMap(showtimeId: String): LiveData<Resource<SeatMapResponse>> {
        val r = MutableLiveData<Resource<SeatMapResponse>>()
        r.value = Resource.loading()
        api.getSeatMap(showtimeId).enqueue(object : Callback<ApiResponse<SeatMapResponse>> {
            override fun onResponse(c: Call<ApiResponse<SeatMapResponse>>, res: Response<ApiResponse<SeatMapResponse>>) {
                if (res.isSuccessful && res.body() != null && res.body()!!.success) {
                    r.value = Resource.success(res.body()!!.data)
                } else {
                    r.value = Resource.error("Tải sơ đồ ghế thất bại")
                }
            }

            override fun onFailure(c: Call<ApiResponse<SeatMapResponse>>, t: Throwable) {
                r.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return r
    }
}
